### PR TITLE
Start video in mpv at the time the video was at in the browser

### LINKED ISSUE #56: Seek the time the video was at in the browser in mpv
**Is your feature request related to a problem? Please describe.**
I sometimes start watching a video in a website, before wanting to continue it in mpv, and sometimes I use timestamped video links, in those cases when I use ff2mpv I would like it to continue in mpv from the same point the video was at in the browser.

**Describe the solution you'd like**
The json message that is sent from the extension to the native host should include another field `time` which would be the time in seconds from the start of the video (the `<video>` tag has a property `currentTime` that supplies it), an example message would look like `{"url":"https://www.youtube.com/watch?v=YE7VzlLtp-4","time":"47.348776"}`. In the python side, you can retrieve it with `message.get("time")` and supply it to mpv with the `--start=time` flag.

**Describe alternatives you've considered**
Seeking the time manually, I'll rather ff2mpv do it for me.

**Additional context**
This is a really small change that I can implement pretty quickly, I want to see if this feature is wanted, if not I'll make it into a fork.

### PR DESCRIPTION
**Pull request by bbb651**
Closes #56

Also I don't fully understand how `browser.tabs.executeScript` works from the mozilla docs, I think currently if a video is playing and the context menu item is used on a link it'll play that url with the time of the video in the current tab, maybe an argument can be added to the `ff2mpv` function to check that.

(I couldn't get the messaging to work between the extention and the python script, and I can't modify the extention because it is signed, so it's currently untested and might not work)

### CODE REVIEW

**Comment by woodruffw** [comment]
Thanks! I'll do a full review of this tonight.

> (I couldn't get the messaging to work between the extention and the python script, and I can't modify the extention because it is signed, so it's currently untested and might not work)

The best way to test it is probably to load it as a temporary extension, which bypasses any integrity checks that AMO imposes. They document the process here: https://extensionworkshop.com/documentation/develop/temporary-installation-in-firefox/

File: ff2mpv.py [hunk 1]
```diff
@@ -11,8 +11,9 @@
 def main():
     message = get_message()
     url = message.get("url")
+    time = message.get("time") or 0
```
Comment (woodruffw) [inline ff2mpv.py:14]:
```suggestion
    time = message.get("time")
```

...then, only inject the `--time=TIME` argument if it's present. Performing it unconditionally might have unintended side effects.

File: ff2mpv.js [hunk 2]
```diff
@@ -2,11 +2,11 @@ function onError(error) {
...
 
-function ff2mpv(url) {
-    browser.tabs.executeScript({
-        code: "video = document.getElementsByTagName('video');video[0].pause();"
-    });
-    browser.runtime.sendNativeMessage("ff2mpv", { url: url }).catch(onError);
+async function ff2mpv(url) {
+    let time = await browser.tabs.executeScript({
+        code: "video = document.getElementsByTagName('video');video[0].pause();video[0].currentTime"
+    }) || 0;
```
Comment (woodruffw) [inline ff2mpv.js:8]: Let's have this fall back to `null` here, per the comment below.

**Comment by woodruffw** [comment]
>  I think currently if a video is playing and the context menu item is used on a link it'll play that url with the time of the video in the current tab, maybe an argument can be added to the `ff2mpv` function to check that.

Ah yeah, this is a problem. You should add a boolean `direct` argument that indicates the "disposition" of the URL we're fetching -- `true` would be a directly selected URL via whatever means, `false` would be the "guess" mode where we just try the page itself.

**Comment by bbb651** [comment]
> The best way to test it is probably to load it as a temporary extension, which bypasses any integrity checks that AMO imposes. They document the process here: https://extensionworkshop.com/documentation/develop/temporary-installation-in-firefox/

I tried that, the extension ran but it failed to communicate with the python script.

I also added a regex to get the time from a timestamped url, I tested the regex with Youtube, Vimeo and Odysee and it works well (in theory it can break in if a site has timestamps in a weird way but I don't think it's an issue, timestamped urls aren't used much).

**Comment by woodruffw** [comment]
> I also added a regex to get the time from a timestamped url, I tested the regex with Youtube, Vimeo and Odysee and it works well (in theory it can break in if a site has timestamps in a weird way but I don't think it's an issue, timestamped urls aren't used much).

This shouldn't be necessary, from my understanding of `youtube-dl` and `mpv` -- `youtube-dl` should be able to parse the timestamp from the URL on its own, and `mpv`'s hook for `youtube-dl` should be smart enough to seek to that timestamp. Were URLs with timestamps not working for you before?

File: ff2mpv.js [hunk 3]
```diff
@@ -2,11 +2,17 @@ function onError(error) {
...
+async function ff2mpv(url, direct) {
+    let time;
+    if (direct) {
+        time = await browser.tabs.executeScript({
+            code: "video = document.getElementsByTagName('video');video[0].pause();video[0].currentTime"
+        });
+    } else {
+        // Try to find time in a url parameter
+        time = parseInt(url.match(/(?<=\W(t|time)=)\d+/)?.[0], 10)
+    }
```
Comment (woodruffw) [inline ff2mpv.js:14]: Per the comment, let's not do this. MPV and `youtube-dl` are smart enough to seek based on the query parameter on their own.

File: ff2mpv.py [hunk 4]
```diff
@@ -11,8 +11,9 @@
 def main():
     message = get_message()
     url = message.get("url")
+    time = message.get("time")
 
-    args = ["mpv", "--no-terminal", "--", url]
+    args = ["mpv", "--no-terminal", f"--start={time}", "--", url] if time else ["mpv", "--no-terminal", "--", url]
```
Comment (woodruffw) [inline ff2mpv.py:16]: Something like this is preferred:

```python
extra_args = []
if time is not None:
    extra_args.append(f"--time={time}")

args = ["mpv", "...", *extra_args, "--", url]
```

Comment (woodruffw) [reply ff2mpv.py:16]: That gives us more flexibility to add other options in the future, with less diff churn.

**Comment by bbb651** [comment]
Sorry it took so long, I haven't been active on github.

**Comment by woodruffw** [comment]
> Sorry it took so long, I haven't been active on github.

Not a problem at all! I don't think it's fair to demand particular deadlines from people contributing to non-professional repos, so you have nothing to apologize for.

I'll do another review pass this weekend.

**Comment by Mayurifag** [comment]
@woodruffw have you done the review? Im very excited and looking forward to see the feature working soon! 👍🏻 ❤️

**Comment by woodruffw** [comment]
Not yet, no.

https://github.com/woodruffw/ff2mpv/pull/57#issuecomment-904212865 still has my main concern with this PR -- `mpv` and `youtube-dl` already know how to coordinate on opening a video at the correct timestamp for most URLs, so any extra JS we inject needs to not interfere with that default behavior and also should not rely on heuristics to guess the right `<video>` element to pull the timestamp from.

So in general, I think this PR needs to do the following in order to get merged:

1. Generalize the current injected JS so that it detects the currently playing `<video>` element, instead of just assuming that the first `<video>` is being played
2. Extract the timestamp from that element
3. Figure out a good way to tell MPV and `youtube-dl` to only use that timestamp if one isn't already present in the URL (we should always treat URLs as authoritative)

**Comment by woodruffw** [comment]
Closing per feedback on the corresponding issue.

**Pull request closed without merging**